src/pint/models/model_builder.py

```diff
@@ -2,317 +2,366 @@
+            # If `timing_model` in param --> component mapping skip
+            # Timing model is the base.
+            if "timing_model" in cps:
                 continue
-            cp = self.select_comp[co]
-            sorted_components.append(cp)
-        return sorted_components
-
-    def search_prefix_param(self, paramList, model, prefix_type):
-        """Check if the Unrecognized parameter has prefix parameter"""
-        prefixs = {}
-        prefix_inModel = model.get_params_of_type_top(prefix_type)
-        for pn in prefix_inModel:
-            par = getattr(model, pn)
-            prefixs[par.prefix] = []
-            for p in paramList:
-                try:
-                    pre, idxstr, idxV = split_prefixed_name(p)
-                    if pre in [par.prefix] + par.prefix_aliases:
-                        prefixs[par.prefix].append(p)
-                except ValueError:  # FIXME: is this meant to catch KeyErrors?
+            # Check if it is a binary component, if yes, skip. It is controlled
+            # by the BINARY tag
+            if len(cps) == 1:  # No conflict, parameter only shows in one component.
+                # Check if it is a binary component, if yes, skip. It is
+                # controlled by the BINARY tag
+                if self.components[cps[0]].category == "pulsar_system":
                     continue
-
-        return prefixs
-
-    def build_model(self, parfile=None, name=""):
-        """Read parfile using the model_instance attribute. Throws error if
-           mismatched coordinate systems detected.
-        Parameters
+                selected_components.add(cps[0])
+                continue
+            # Has conflict, same parameter shows in different components
+            # Only record the conflict here and do nothing, if there is any
+            # component unique parameter show in the parfile, the component will
+            # be selected.
+            if len(cps) > 1:
+                # Add conflict to the conflict graph
+                for cp in cps:
+                    temp_cf_cp = copy.deepcopy(cps)
+                    temp_cf_cp.remove(cp)
+                    conflict_components[cp].update(set(temp_cf_cp))
+                continue
+        # Check if the selected component in the confilict graph. If it is
+        # remove the selected componens with its conflict components.
+        for ps_cp in selected_components:
+            cf_cps = conflict_components.get(ps_cp, None)
+            if cf_cps is not None:  # Had conflict, but resolved.
+                for cf_cp in cf_cps:
+                    del conflict_components[cf_cp]
+                del conflict_components[ps_cp]
+        return selected_components, conflict_components, unrec_param
+
+    def _add_indexed_params(self, timing_model, indexed_params):
+        """Add the parameters with unknown number/indexed in parfile (maskParameter/
+        prefixParameter) to timing model.
```

Comment:
First line of a docstring should be on one line. If too long, write a one-line summary and put this in the first paragraph.